File: cpp/src/ch05/thunder_search.cc
#include <assert.h>
#include "thunder_search.h"
#include "time_keeper.h"

namespace thunder
{
    double Node::_increment(double value)
    {
        this->w_ += value;
        ++this->n_;
        return value;
    }

    double Node::evaluate()
    {
        if (this->state_.is_done())
        {
            double value = this->state_.teban_score();
            return this->_increment(value);
        }

        if (this->child_nodes_.empty())
        {
            double value = this->state_.get_score_rate();
            this->_increment(value); // expand が入るから return しない
            this->expand();
            return value;
        }

        double value = 1. - this->next_child_node().evaluate();
        return this->_increment(value);
    }

    Node &Node::next_child_node()
    {
        for (auto &child_node : this->child_nodes_)
        {
            if (child_node.n_ == 0)
                return child_node;
        }

        double best_value = -INF;
        int best_action_index = -1;
        for (int i = 0; i < (int)this->child_nodes_.size(); i++)
        {
            const auto &child_node = this->child_nodes_[i];
            double thunder_value = 1. - child_node.w_ / child_node.n_;
            if (thunder_value > best_value)
            {
                best_action_index = i;
                best_value = thunder_value;
            }
        }
        return this->child_nodes_[best_action_index];
    }

    int Node::best_action()
    {
        auto legal_actions = this->state_.legal_actions();
        // return argmax of child_node.n
        int best_action_index = -1;
        int best_score = -1;
        assert(legal_actions.size() == this->child_nodes_.size());
        for (int i = 0; i < (int)legal_actions.size(); i++)
        {
            int n = this->child_nodes_[i].n_;
            if (n > best_score)
            {
                best_action_index = i;
                best_score = n;
            }
        }
        return legal_actions[best_action_index];
    }

    void Node::expand()
    {
        auto legal_actions = this->state_.legal_actions();
        this->child_nodes_.clear();
        for (const auto action : legal_actions)
        {
            this->child_nodes_.emplace_back(this->state_);
            this->child_nodes_.back().state_.advance(action);
        }
    }

    int thunder_search_action(const State &state,
                              const int playout_number)
    {
        Node root_node = Node(state);
        root_node.expand();
        for (int i = 0; i < playout_number; i++)
            root_node.evaluate();

        return root_node.best_action();
    }

    int thunder_search_action_with_timekeeper(
        const State &state,
        const int64_t time_threshold)
    {
        Node root_node = Node(state);
        root_node.expand();

        auto time_keeper = TimeKeeper(time_threshold);
        while (!time_keeper.is_time_over())
            root_node.evaluate();

        return root_node.best_action();
    }
}

File: python/src/thunder_book/ch05/thunder_search.py
from __future__ import annotations

import fire
import numpy as np

from thunder_book.ch05.average_score import average_score
from thunder_book.ch05.iterative_deepening import iterative_deepening_action
from thunder_book.ch05.maze_state import AlternateMazeState as State
from thunder_book.ch05.maze_state import MazeParams, MCTSParams
from thunder_book.ch05.monte_carlo_tree_search import (
    mcts_action,
    mcts_action_with_time_threshold,
)
from thunder_book.ch05.time_keeper import TimeKeeper
from thunder_book.util import setup_logging


class TNode:
    def __init__(self, state: State) -> None:
        """
        node for thunder search
        use "TNode" to avoid name conflict with "Node" in monte_carlo_tree_search.py
        """
        self.state: State = state.copy()
        self.child_nodes: list[TNode] = []
        self.n = 0
        self.w = 0

    def __repr__(self) -> str:
        return f"<TNode: n({self.n}), w({self.w:.1f})>"

    def expand(self) -> None:
        legal_actions = self.state.legal_actions()
        self.child_nodes = []
        for action in legal_actions:
            self.child_nodes.append(TNode(self.state))
            self.child_nodes[-1].state.advance(action)

    def _increment(self, value: float) -> None:
        self.w += value
        self.n += 1

    def evaluate(self) -> float:
        # todo: これ explore に名前を変えたい
        if self.state.is_done():
            value = self.state.teban_score()
            self._increment(value)
            return value

        # no child nodes
        if not self.child_nodes:
            value = self.state.get_score_rate()
            self._increment(value)
            self.expand()
            return value

        # base case, has child nodes
        value = 1 - self.next_child_node().evaluate()
        self._increment(value)
        return value

    def next_child_node(self) -> TNode:
        # select child node that has not been evaluated
        for child_node in self.child_nodes:
            if child_node.n == 0:
                return child_node

        # select child node with the highest thunder_value
        thunder_values = [(1 - c.w / c.n) for c in self.child_nodes]
        best_index = np.argmax(thunder_values)
        return self.child_nodes[best_index]

    def best_action(self) -> int:
        legal_actions = self.state.legal_actions()
        assert len(legal_actions) == len(self.child_nodes)
        # select node with the highest n
        best_index = np.argmax([c.n for c in self.child_nodes])
        return legal_actions[best_index]

    def print_tree(self, depth: int = 0, action=None) -> str:
        ss = "__ " * depth
        ss += f"{action}=>" if action is not None else ""
        ss += f"{self.n}({self.w:.1f})\n"
        for action, child_node in zip(self.state.legal_actions(), self.child_nodes):
            ss += child_node.print_tree(depth + 1, action)
        return ss

    def __str__(self) -> str:
        return self.print_tree()


def thunder_search_action(state: State, playout_number: int):
    node = TNode(state)
    node.expand()
    for _ in range(playout_number):
        # print(node.state)
        # breakpoint()
        node.evaluate()
    return node.best_action()


def thunder_search_action_with_time_threshold(state: State, time_threshold: int):
    root_node = TNode(state)
    root_node.expand()
    time_keeper = TimeKeeper(time_threshold)
    while not time_keeper.is_time_over():
        root_node.evaluate()
    return root_node.best_action()


def thunder_search_vs_mcts(num_playouts=100):
    num_games = 100
    thunder_search_f = lambda x: thunder_search_action(x, num_playouts)
    mcts_action_f = lambda x: mcts_action(
        x,
        num_playouts,
        mcts_params=MCTSParams(
            c=1.0,
            expand_threshold=10,
        ),
    )

    actions_wb = (thunder_search_f, mcts_action_f)
    params = MazeParams(width=5, height=5, end_turn=10)
    win_rate = average_score(num_games, actions_wb, params=params)

    print(f"thunder_search vs mcts: {win_rate:.2f} in {num_playouts=}, {num_games=}")


def thunder_vs_mcts_timebound(time_threshold=1, num_games=100):
    thunder_search_f = lambda x: thunder_search_action_with_time_threshold(x, time_threshold)
    mcts_action_f = lambda x: mcts_action_with_time_threshold(
        x,
        time_threshold,
        mcst_params=MCTSParams(c=1.0, expand_threshold=10),
    )
    actions_wb = (thunder_search_f, mcts_action_f)
    params = MazeParams(width=5, height=5, end_turn=10)
    win_rate = average_score(num_games, actions_wb, params=params)

    print(f"thunder_search vs mcts timebound: {win_rate:.2f} in {time_threshold=}, {num_games=}")


def thunder_vs_iterative_deepening_timebound(time_threshold=1, num_games=100):
    thunder_search_f = lambda x: thunder_search_action_with_time_threshold(x, time_threshold)
    iterative_deepening_action_f = lambda x: iterative_deepening_action(x, time_threshold)
    actions_wb = (thunder_search_f, iterative_deepening_action_f)
    params = MazeParams(width=5, height=5, end_turn=10)
    win_rate = average_score(num_games, actions_wb, params=params)

    print(
        f"thunder_search vs iterative deepening timebound: {win_rate:.2f} in {time_threshold=}, {num_games=}"
    )


def play_one():
    params = MazeParams(width=5, height=5, end_turn=10)
    state = State(0, params=params)
    print(state)
    # breakpoint()
    action = thunder_search_action(state, 300)
    print(action)


def main(game="one", *args, **kwargs):
    if game == "one":
        return play_one()
    if game == "time":
        return thunder_vs_mcts_timebound(*args, **kwargs)
    if game == "compare":
        return thunder_search_vs_mcts(*args, **kwargs)
    if game == "alpha_beta":
        return thunder_vs_iterative_deepening_timebound(*args, **kwargs)


if __name__ == "__main__":
    setup_logging()
    fire.Fire(main)
